### PR TITLE
Extends SoftAssertionsExtension to injection of constructor parameters.

### PR DESCRIPTION
**Pull request by fduminy**
#### Check List:
* Extends SoftAssertionsExtension to injection of constructor parameters
* Unit tests : YES
* Javadoc with a code example (on API only) : YES

### CODE REVIEW

**Comment by joel-costigliola** [comment]
@fduminy can you explain with an example the problem yo are trying to solve with this PR?

**Comment by fduminy** [comment]
@joel-costigliola here is an example : 

```java
ExtendWith(SoftAssertionsExtension.class)
 class ExampleTestCase {
    private final SoftAssertions softly;

    ExampleTestCase(SoftAssertions softly) {
      this.softly = softly;
    }

    @Test
    void multipleFailures() {
       softly.assertThat(2 * 3).isEqualTo(0);
       softly.assertThat(Arrays.asList(1, 2)).containsOnly(1);
       softly.assertThat(1 + 1).isEqualTo(2);
    }
}
```

It can be particularly useful when there are many test methods using SoftAssertions in the same class.

**Comment by joel-costigliola** [comment]
We already have support for JUnit 5 https://assertj.github.io/doc/#assertj-core-junit5-soft-assertions, I'm not sure to see what is missing with the current support, can clarify what is missing?

**Comment by fduminy** [comment]
The link you gave doesn't show an example with injection in constructor. I don't think it's currently supported by assertj.

I think we should support different styles of junit 5 tests to give users more flexibility instead of forcing them to use only method injection style. Moreover, migration from junit 4 rule is easier with constructor injection since it avoids changing each test method.

**Comment by joel-costigliola** [comment]
Ok that makes sense now

**Comment by joel-costigliola** [comment]
@fduminy what about taking https://github.com/joel-costigliola/assertj-core/pull/1784 approach? WDTY?

**Comment by fduminy** [comment]
I replyed in #1784

**Comment by joel-costigliola** [comment]
@fduminy as in https://github.com/joel-costigliola/assertj-core/pull/1784#issuecomment-593849219 I think there is a fundamental issue: if we run tests in parallel, the soft assertions field is going to be reset while test instances are being initialized, leading to mix soft assertions errors collected.

Do you see away around that?

**Comment by filiphr** [comment]
JUnit creates a new test instance for every test (unless `Lifecycle.PER_CLASS` is used). According to [Parallel Execution](https://junit.org/junit5/docs/current/user-guide/#writing-tests-parallel-execution), even if you enable parallelization per test node (parallel for the same class) you still have to explicitly mark the class or test method with `@Execution(CONCURRENT)`.

I think that for a constructor injection, or even for #1784 it should be OK to inject the field

File: src/main/java/org/assertj/core/api/ISoftAssertions.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * Copyright 2012-2020 the original author or authors.
+ */
+package org.assertj.core.api;
+
+public interface ISoftAssertions {
```
Comment (dstekanov) [inline src/main/java/org/assertj/core/api/ISoftAssertions.java]: maybe it is worth to add javadoc here

**Comment by joel-costigliola** [comment]
@filiphr thanks for the comment.

**Comment by kriegfrj** [comment]
> @filiphr thanks for the comment.
> 
> @sbrannen do you think as @filiphr suggested that we can use a `SoftAssertions` field initialized with constructor injection? (except for classes combining `Lifecycle.PER_CLASS` with `@Execution(CONCURRENT)`)

Actually, I think that constructor injection would be a problem for `SoftAssertions` if you're using `PER_CLASS`, regardless of whether it's single threaded or concurrent. That's because in `PER_CLASS` the constructor is only called once, so you don't get a fresh `SoftAssertions` object for each test.

I don't think that this should be a reason not to support constructor injection though (more on that in https://github.com/joel-costigliola/assertj-core/issues/1824#issuecomment-608992199). But to avoid nasty surprises, we should do a sanity check - `supportsParameter()` should throw a `ParameterResolutionException` if it notices that we're trying to inject a soft assertions object into a constructor and ExtensionContext.getTestInstanceLifecycle() is `PER_CLASS`.

**Comment by joel-costigliola** [comment]
@fduminy I have just pushed a [commit](https://github.com/joel-costigliola/assertj-core/commit/a4c847cd49b0537702b7756389d63f9d526993b8) with a refactored `SoftlyExtension` approach without annotating fields, I'm not sure there is a lot of value doing another one for constructor, WDYT?

File: src/main/java/org/assertj/core/api/ISoftAssertions.java (hunk 1 above)
Comment (kriegfrj) [reply src/main/java/org/assertj/core/api/ISoftAssertions.java:15]: `ISoftAssertions` is similar in concept and scope to `SoftAssertionsProvider`, which was merged in #1817. I think that at least this part of this PR is redundant.

**Comment by joel-costigliola** [comment]
I'm closing this issue as we now support field injection, even though it's not constructor injection per se, I believe it solves the same issue.
Please comment if you disagree.

**Pull request closed without merging**